File: server/src/chat/token-estimator.ts
/**
 * 粗略估算 Token 消耗，用于成本控制和上下文长度规避。
 * 并非精确的 BPE 分词，通过字符权重实现平衡。
 */
import type OpenAI from 'openai';

type Message = OpenAI.Chat.Completions.ChatCompletionMessageParam;

const CJK_CHARS_PER_TOKEN = 1.2; // 中文字符平均 Token 消耗
const ASCII_CHARS_PER_TOKEN = 4; // 英文字符/符号平均 Token 消耗
const MESSAGE_OVERHEAD = 4;      // 每条消息的基础 Overhead (role, name 等)

// 覆盖常用中日韩 Unicode 字符范围
const CJK_RANGE = /[\u4e00-\u9fff\u3400-\u4dbf\uf900-\ufaff\u{20000}-\u{2a6df}\u{2a700}-\u{2b73f}]/u;

export function estimateTokens(text: string): number {
	let cjkChars = 0;
	let asciiChars = 0;
	for (const ch of text) {
		if (CJK_RANGE.test(ch)) cjkChars++;
		else asciiChars++;
	}
	return Math.ceil(cjkChars / CJK_CHARS_PER_TOKEN + asciiChars / ASCII_CHARS_PER_TOKEN);
}

export function estimateMessageTokens(msg: Message): number {
	if (typeof msg.content === 'string') {
		return MESSAGE_OVERHEAD + estimateTokens(msg.content);
	}
	return MESSAGE_OVERHEAD + estimateTokens(JSON.stringify(msg));
}

export function estimateMessagesTokens(msgs: Message[]): number {
	return msgs.reduce((sum, m) => sum + estimateMessageTokens(m), 0);
}

File: py-server/src/chat/token_estimator.py
"""粗略估算 Token，用于上下文预算控制。

并非精确 BPE 分词，通过中文/ASCII 字符权重实现。
"""

import json
import re

CJK_CHARS_PER_TOKEN = 1.2
ASCII_CHARS_PER_TOKEN = 4
MESSAGE_OVERHEAD = 4

# 覆盖常用中日韩 Unicode 字符范围
_CJK_RE = re.compile(
    r"[\u4e00-\u9fff\u3400-\u4dbf\uf900-\ufaff\U00020000-\U0002a6df\U0002a700-\U0002b73f]"
)


def estimate_tokens(text: str) -> int:
    if not text:
        return 0
    cjk = len(_CJK_RE.findall(text))
    ascii_count = len(text) - cjk
    import math

    return math.ceil(cjk / CJK_CHARS_PER_TOKEN + ascii_count / ASCII_CHARS_PER_TOKEN)


def estimate_message_tokens(msg: dict) -> int:
    content = msg.get("content")
    if isinstance(content, str):
        return MESSAGE_OVERHEAD + estimate_tokens(content)
    return MESSAGE_OVERHEAD + estimate_tokens(
        json.dumps(msg, ensure_ascii=False, default=str)
    )


def estimate_messages_tokens(msgs: list[dict]) -> int:
    return sum(estimate_message_tokens(m) for m in msgs)
